Load CSV File in Google Colab

In [ ]:
# Install necessary libraries (if not already installed)
!pip install pandas numpy openai transformers

In [ ]:
import os
from google.colab import drive

# Mount Google Drive
drive.mount('/content/drive')

# List files in your Google Drive directory
os.listdir('/content/drive/My Drive')


Mounted at /content/drive


['COMP971 Dissertation N Smith.pdf',
 'RL_DA.pdf',
 'paper#2.pdf',
 'Colab Notebooks',
 'parsed_clustered_recipe_with_difficulty.csv',
 'offload',
 'Git']

In [ ]:
# Import required libraries
import pandas as pd

# Define the file path (adjust if the file is in a subdirectory)
file_path = '/content/drive/My Drive/parsed_clustered_recipe_with_difficulty.csv'

# Load the CSV file into a pandas DataFrame
df = pd.read_csv(file_path)

# Display the first few rows of the dataset
print("Dataset loaded successfully!")
df.head()



Dataset loaded successfully!


title  \
0                                           Gazpacho   
1  Rice Paper Wrapped Halibut Roasted with Foie G...   
2  Lemon Cupcakes with Blueberry Compote Filling ...   
3  Soft Pretzels with Queso Poblano Sauce and Mus...   
4                               Sweet Biscuit Wreath   

                                         ingredients  \
0  ['8 tomatoes, quartered', 'Kosher salt', '1 re...   
1  ['4 large, round rice wrappers', '4 (6-ounce) ...   
2  ['1 1/2 cups granulated sugar', '1 stick (1/2 ...   
3  ['1 tablespoon unsalted butter', '1 tablespoon...   
4  ['1/2 cup sugar', '1 tablespoon pumpkin pie sp...   

                                          directions  \
0  ['["Add the tomatoes to a food processor with ...   
1  ['["Spread foie mousse on each fillet and top ...   
2  ['["For the cupcakes: Preheat the oven to 350 ...   
3  ['["Melt the butter in a medium saucepan over ...   
4  ['["Preheat oven to 350 degrees F', '", "On a ...   

                                                link     source  \
0         www.foodnetwork.com/recipes/gazpacho1.html  Recipes1M   
1  www.foodnetwork.com/recipes/rice-paper-wrapped...  Recipes1M   
2  www.foodnetwork.com/recipes/lemon-cupcakes-wit...  Recipes1M   
3  www.foodnetwork.com/recipes/bobby-flay/soft-pr...  Recipes1M   
4  www.foodnetwork.com/recipes/sandra-lee/sweet-b...  Recipes1M   

                                                 NER                 site  \
0  ["yellow bell pepper", "extra-virgin olive oil...  www.foodnetwork.com   
1  ["egg", "gras", "ginger", "heavy cream", "hone...  www.foodnetwork.com   
2  ["buttermilk", "sugar", "baking powder", "vani...  www.foodnetwork.com   
3  ["dijon mustard", "salt", "active dry yeast", ...  www.foodnetwork.com   
4  ["egg", "sugar", "powdered sugar", "water", "m...  www.foodnetwork.com   

                            preprocessed_ingredients  \
0  fine dice pepper kosher chop extra salt bell s...   
1  fine dice pepper long garnish deseed truffl ho...   
2  extract soda blueberri vanilla flour purpos sa...   
3  pepper soda brown activ kosher monterey jack h...   
4  cinnamon spice egg ground ounc water refriger ...   

                             preprocessed_directions  \
0  pepper garnish chill food add chop combin salt...   
1  pinot brown garnish vineyard honey sauc demi d...   
2  food blueberri mixtur insert begin refriger ju...   
3  boil cover long brown equal honey slightli she...   
4  cutter cover bottom cinnamon move roll sprinkl...   

                                         full_recipe  ... cluster_label  \
0  ["8 tomatoes, quartered", "Kosher salt", "1 re...  ...             5   
1  ["4 large, round rice wrappers", "4 (6-ounce) ...  ...             3   
2  ["1 1/2 cups granulated sugar", "1 stick (1/2 ...  ...             0   
3  ["1 tablespoon unsalted butter", "1 tablespoon...  ...             3   
4  ["1/2 cup sugar", "1 tablespoon pumpkin pie sp...  ...             5   

  most_similar_recipe_index combined_most_similar_recipe cluster  \
0                        34                         34.0       1   
1                        17                         34.0       1   
2                        70                         34.0       1   
3                        42                         34.0       1   
4                        70                         34.0       1   

                                          parsed_ner ingredient_count  \
0  ['extra-virgin olive oil', 'yellow bell pepper...                9   
1  ['truffle oil', 'ginger', 'heavy cream', 'chic...               19   
2  ['vanilla', 'cream cheese', 'egg whites', 'pow...               17   
3  ['water', 'whole-grain mustard', 'cilantro', '...               21   
4  ['water', 'powdered sugar', 'pumpkin pie spice...                8   

   step_count rare_ingredients_count  preparation_keywords  \
0           0                      0                     0   
1           0                      0                     1   
2          

In [ ]:
# Check dataset info
df.info()

# Check for missing values
print("Missing values in each column:")
print(df.isnull().sum())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49441 entries, 0 to 49440
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   title                         49441 non-null  object 
 1   ingredients                   49441 non-null  object 
 2   directions                    49441 non-null  object 
 3   link                          49441 non-null  object 
 4   source                        49441 non-null  object 
 5   NER                           49441 non-null  object 
 6   site                          49441 non-null  object 
 7   preprocessed_ingredients      49439 non-null  object 
 8   preprocessed_directions       49440 non-null  object 
 9   full_recipe                   49441 non-null  object 
 10  preprocessed_full_recipe      49439 non-null  object 
 11  quantities                    49441 non-null  object 
 12  units                         49441 non-null  object 
 13  c

In [ ]:
# Drop rows with missing values (if these columns are critical)
df_cleaned = df.dropna(subset=['preprocessed_ingredients', 'preprocessed_directions', 'preprocessed_full_recipe'])

# Alternatively, fill missing values with an empty string
df_filled = df.fillna('')

# Confirm no missing values remain
print("Missing values after cleaning:")
print(df_filled.isnull().sum())

In [ ]:
#!pip install sentence-transformers



In [ ]:
# Display all column names
print(df.columns)


Index(['title', 'ingredients', 'directions', 'link', 'source', 'NER', 'site',
       'preprocessed_ingredients', 'preprocessed_directions', 'full_recipe',
       'preprocessed_full_recipe', 'quantities', 'units',
       'cleaned_ingredients', 'standardized_ingredients', 'averaged_embedding',
       'duration', 'most_similar_recipes', 'Step_Complexity',
       'Technique_Diversity', 'Ingredient_Diversity', 'Total_Complexity',
       'Difficulty_Level', 'cluster_label', 'most_similar_recipe_index',
       'combined_most_similar_recipe', 'cluster', 'parsed_ner',
       'ingredient_count', 'step_count', 'rare_ingredients_count',
       'preparation_keywords', 'new_difficulty_score'],
      dtype='object')


In [ ]:
import pandas as pd
from google.colab import drive

# Mount Google Drive to access the dataset
drive.mount('/content/drive')

# Load the dataset from Google Drive
file_path = '/content/drive/My Drive/parsed_clustered_recipe_with_difficulty.csv'
df_cleaned = pd.read_csv(file_path)

# Fill missing values in the target column
df_cleaned['preprocessed_full_recipe'] = df_cleaned['preprocessed_full_recipe'].fillna('')

# Verify the dataset
print("Dataset loaded successfully. First few rows:")
print(df_cleaned[['title', 'preprocessed_full_recipe']].head())




Mounted at /content/drive
Dataset loaded successfully. First few rows:
                                               title  \
0                                           Gazpacho   
1  Rice Paper Wrapped Halibut Roasted with Foie G...   
2  Lemon Cupcakes with Blueberry Compote Filling ...   
3  Soft Pretzels with Queso Poblano Sauce and Mus...   
4                               Sweet Biscuit Wreath   

                            preprocessed_full_recipe  
0  fine dice pepper kosher chop extra salt bell s...  
1  fine dice pepper long garnish deseed truffl ho...  
2  extract soda blueberri vanilla flour purpos sa...  
3  pepper soda brown activ kosher monterey jack h...  
4  cinnamon spice egg ground ounc water refriger ...  


In [ ]:
from sentence_transformers import SentenceTransformer
import numpy as np
import pandas as pd
from google.colab import drive

# Mount Google Drive to access the dataset
drive.mount('/content/drive')

# Load the dataset
file_path = '/content/drive/My Drive/parsed_clustered_recipe_with_difficulty.csv'
df_cleaned = pd.read_csv(file_path)

# Fill missing values in the target column
df_cleaned['preprocessed_full_recipe'] = df_cleaned['preprocessed_full_recipe'].fillna('')

# Load the SentenceTransformer model
model = SentenceTransformer('all-MiniLM-L6-v2')

# Extract the text column for embeddings
sentences = df_cleaned['preprocessed_full_recipe'].tolist()

# Generate embeddings
print("Generating embeddings...")
embeddings = model.encode(sentences, batch_size=32, show_progress_bar=True)

# Add embeddings as a new column
df_cleaned['generated_embedding'] = embeddings.tolist()
print("Embeddings generated and added to the DataFrame.")

# Save the updated dataset to Google Drive
csv_path = '/content/drive/My Drive/updated_recipes_with_generated_embeddings.csv'
df_cleaned.to_csv(csv_path, index=False)
print(f"Updated dataset saved to: {csv_path}")

# Save only the embeddings as a .npy file
npy_path = '/content/drive/My Drive/recipe_embeddings.npy'
np.save(npy_path, embeddings)
print(f"Embeddings saved to: {npy_path}")





Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Generating embeddings...


Batches:   0%|          | 0/1546 [00:00<?, ?it/s]

Embeddings generated and added to the DataFrame.
Updated dataset saved to: /content/drive/My Drive/updated_recipes_with_generated_embeddings.csv
Embeddings saved to: /content/drive/My Drive/recipe_embeddings.npy


In [ ]:
import pandas as pd
import numpy as np

# Load the updated dataset
csv_path = '/content/drive/My Drive/updated_recipes_with_generated_embeddings.csv'
df_cleaned = pd.read_csv(csv_path)

# Load the standalone embeddings
npy_path = '/content/drive/My Drive/recipe_embeddings.npy'
embeddings = np.load(npy_path)

# Verify the dataset
print("Dataset sample:")
print(df_cleaned.head())

# Verify the embeddings
print(f"Embeddings shape: {embeddings.shape}")
print(f"Sample embedding (first recipe): {embeddings[0]}")


Dataset sample:
                                               title  \
0                                           Gazpacho   
1  Rice Paper Wrapped Halibut Roasted with Foie G...   
2  Lemon Cupcakes with Blueberry Compote Filling ...   
3  Soft Pretzels with Queso Poblano Sauce and Mus...   
4                               Sweet Biscuit Wreath   

                                         ingredients  \
0  ['8 tomatoes, quartered', 'Kosher salt', '1 re...   
1  ['4 large, round rice wrappers', '4 (6-ounce) ...   
2  ['1 1/2 cups granulated sugar', '1 stick (1/2 ...   
3  ['1 tablespoon unsalted butter', '1 tablespoon...   
4  ['1/2 cup sugar', '1 tablespoon pumpkin pie sp...   

                                          directions  \
0  ['["Add the tomatoes to a food processor with ...   
1  ['["Spread foie mousse on each fillet and top ...   
2  ['["For the cupcakes: Preheat the oven to 350 ...   
3  ['["Melt the butter in a medium saucepan over ...   
4  ['["Preheat oven to 350 deg

In [ ]:
import pandas as pd
import numpy as np

from google.colab import drive

# Mount Google Drive to access the dataset
drive.mount('/content/drive')
# Reload the dataset
csv_path = '/content/drive/My Drive/updated_recipes_with_generated_embeddings.csv'
df_cleaned = pd.read_csv(csv_path)

# Reload the embeddings
npy_path = '/content/drive/My Drive/recipe_embeddings.npy'
embeddings = np.load(npy_path)

# Verify reload
print(f"Dataset rows: {len(df_cleaned)}, Embedding shape: {embeddings.shape}")
print(f"Sample embedding: {embeddings[0]}")


Mounted at /content/drive
Dataset rows: 49441, Embedding shape: (49441, 384)
Sample embedding: [-9.64803398e-02 -8.45913216e-03  2.68462542e-02  2.69799884e-02
 -5.04813977e-02  4.49861102e-02  1.94809344e-02 -2.54545957e-02
 -2.51530707e-02 -7.12438226e-02  5.73005788e-02 -1.01706341e-01
 -3.98705788e-02 -7.36513957e-02 -1.11441612e-02 -8.88769031e-02
  1.06734268e-01  5.51453605e-02 -1.35981008e-01 -1.26022294e-01
  1.03040747e-02 -1.10851619e-02  3.66344675e-02 -4.16697226e-02
  1.46005419e-03  6.49724081e-02  2.85455957e-02  4.68299575e-02
  2.99233068e-02 -1.52851924e-01  2.37356815e-02  5.81064634e-02
  3.35335173e-02 -2.52162851e-02  9.35671013e-03 -3.06841563e-02
  1.31243486e-02 -6.15646206e-02  3.90981175e-02 -4.17697988e-03
  3.33511606e-02  3.31576774e-03  2.24213433e-02  3.68365459e-02
 -3.11797354e-02 -9.53553244e-03 -7.02936500e-02 -1.36744874e-02
  3.92564125e-02 -3.95387523e-02 -5.22357076e-02 -3.92927192e-02
 -3.05561814e-02  7.17461435e-03  1.83901619e-02 -2.03214120

Build a Simple Recommendation System: Use filtering or embeddings for ingredient-based recipe recommendations.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
from sentence_transformers import SentenceTransformer

# Load the same model used for recipe embeddings
model = SentenceTransformer('all-MiniLM-L6-v2')

def get_query_embedding(ingredients):
    """
    Generate an embedding for the user input ingredients.
    """
    query_embedding = model.encode(ingredients, convert_to_numpy=True)
    return query_embedding





/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Calculate Similarity and Recommend Recipes

In [ ]:
def recommend_recipes(user_ingredients, embeddings, df_cleaned, top_n=5):
    """
    Recommend recipes based on user ingredients.

    Args:
    - user_ingredients: String containing user ingredients (e.g., "chicken, garlic, onion").
    - embeddings: Precomputed embeddings for recipes.
    - df_cleaned: DataFrame containing recipe details.
    - top_n: Number of recommendations to return.

    Returns:
    - DataFrame with recommended recipes and their similarity scores.
    """
    # Generate query embedding for user ingredients
    query_embedding = get_query_embedding(user_ingredients)

    # Calculate cosine similarity between query and all recipe embeddings
    similarities = cosine_similarity([query_embedding], embeddings)[0]

    # Get indices of top N most similar recipes
    top_indices = similarities.argsort()[-top_n:][::-1]  # Sort and reverse

    # Extract recommended recipes
    recommended_recipes = df_cleaned.iloc[top_indices].copy()
    recommended_recipes['similarity_score'] = similarities[top_indices]

    return recommended_recipes[['title', 'ingredients', 'directions', 'similarity_score']]



Test the Recommendation System
Provide a sample input (e.g., user-selected ingredients) and test the function.

In [ ]:
# Example: User input ingredients
user_ingredients = "chicken, garlic, onion"

# Get recommendations
recommended = recommend_recipes(user_ingredients, embeddings, df_cleaned, top_n=5)

# Display recommendations
print("Recommended Recipes:")
print(recommended)


Recommended Recipes:
                                 title  \
9888                      Jolly Greens   
2781     Monday to Friday Turkey Gumbo   
48944  Picnic Potato and Chicken Salad   
1982          Tropical Chicken Patties   
46450                  Shrimp Cocktail   

                                             ingredients  \
9888   ['4 ham hocks', '2 yellow onions', '3 cloves g...   
2781   ['1/4 cup olive oil', '2 scallions, chopped', ...   
48944  ['4 Idaho potatoes, peeled and cubed', '1 1/2 ...   
1982   ['1 1/4 pounds ground chicken', '2 small jalap...   
46450  ['6 plum tomatoes, diced', '1 green bell peppe...   

                                              directions  similarity_score  
9888   ['["Cook ham hocks, whole onions, whole garlic...          0.747949  
2781   ['["Over high heat olive oil in a large saucep...          0.726300  
48944  ['["Preheat oven to 400 degrees F', '", "Place...          0.725316  
1982   ['["Put the chicken, half of the green jalapen... 

Explanation of the Output
Recommended Recipes:

Titles like "Jolly Greens" and "Monday to Friday Turkey Gumbo" are returned as top matches.
These recipes are semantically similar to the input ingredients ("chicken, garlic, onion").
Ingredients:

Each recipe’s ingredient list is displayed, which aligns with the user’s query in terms of semantic similarity.
Directions:

The cooking instructions provide a complete recipe workflow for the user.
Similarity Scores:

Scores (e.g., 0.747949) indicate how closely each recipe matches the user’s input.
Higher scores = better match.

What Works Well
Semantic Matching: The system uses embeddings to find semantically similar recipes rather than relying on exact ingredient matches. For example, recipes with "chicken" and "garlic" but different preparation styles are included.
Flexibility: The system adapts to user inputs with diverse wording or ingredient lists.


Integrate an LLM for Recipe Suggestions and Customisation
Use OpenAI’s GPT for dynamic responses.

Overview
Goal: Add a natural language interface to your recipe recommendation system, allowing users to:
 - Get recipe suggestions.
Customise recipes (e.g., "Make it vegan" or "Add garlic").
 - Implementation:
Use an LLM to process user queries and interact with the recommendation system or dataset.
Free Option: Use Hugging Face’s transformers library with a free model like tiiuae/falcon-7b-instruct (instruction-tuned for Q&A).

In [ ]:
#!pip install transformers



In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the Falcon-7B-Instruct model from Hugging Face
model_name = "tiiuae/falcon-7b-instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

print("Model and tokenizer loaded successfully!")


tokenizer_config.json:   0%|          | 0.00/1.13k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.73M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/281 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.05k [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/17.7k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.95G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/4.48G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/117 [00:00<?, ?B/s]

Model and tokenizer loaded successfully!


Define a Function to Generate Responses
This function will take a user query (e.g., "Make the recipe vegan") and use the LLM to generate a response.

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the Falcon-7B-Instruct model from Hugging Face
model_name = "tiiuae/falcon-7b-instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Add padding token to the tokenizer
tokenizer.pad_token = tokenizer.eos_token
# or tokenizer.add_special_tokens({'pad_token': '[PAD]'}) # alternative

print("Model and tokenizer loaded successfully!")

def get_recipe_customisation(user_query, context, max_new_tokens=100):
    """
    Generate a recipe suggestion or customisation based on user input.

    Args:
    - user_query: User's customisation request or question.
    - context: Contextual information (e.g., recipe details).
    - max_new_tokens: Number of new tokens to generate.

    Returns:
    - Generated response from the LLM.
    """
    # Combine the user query with the recipe context
    input_text = f"Context:\n{context}\n\nUser Query:\n{user_query}\n\nResponse:"

    # Tokenize input with padding and truncation
    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512  # Adjust max length for tokenized input
    )

    # Generate response
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],  # Add attention mask
        max_new_tokens=max_new_tokens,  # Specify new token limit
        pad_token_id=tokenizer.eos_token_id  # Set pad token to eos token
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    return response



/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/1.13k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.73M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/281 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.05k [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/17.7k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.95G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/4.48G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/117 [00:00<?, ?B/s]

Model and tokenizer loaded successfully!


Use the LLM with Recipe Data
Test the integration by customising a recipe from your dataset.

In [ ]:
# Example: User selects a recipe
recipe_index = 0  # Index of the selected recipe
recipe_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                 f"Ingredients: {df_cleaned['ingredients'][recipe_index]}\n" \
                 f"Directions: {df_cleaned['directions'][recipe_index]}"

# Example user query
user_query = "Make this recipe vegan and suggest a side dish."

# Get the customised recipe suggestion
customised_response = get_recipe_customisation(user_query, recipe_context, max_new_tokens=200)


print("Customised Recipe Response:")
print(customised_response)



Customised Recipe Response:
Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, cut into small dice', '1 red bell pepper, cut into small dice', '1 yellow bell pepper, cut into small dice', '1/2 cucumber, cut into small dice', 'Extra-virgin olive oil, for drizzling', '3 leaves fresh basil, finely chopped']
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine the onions, bell peppers and cucumbers with the tomato puree in a large bowl', '", "Chill at least 1 hour', '", "Drizzle with olive oil, garnish with chopped basil and serve', '"]']

User Query:
Make this recipe vegan and suggest a side dish.

Response:
<pre><code>[['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth", "Combine the onions, bell peppers and cucumbers with the tomato puree in a large bowl", "Chill at least 1 hour", "Drizzle with olive oil, garni

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the Falcon-7B-Instruct model from Hugging Face
model_name = "tiiuae/falcon-7b-instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Add padding token to the tokenizer
tokenizer.pad_token = tokenizer.eos_token

print("Model and tokenizer loaded successfully!")

# Function to generate recipe customisation
def get_recipe_customisation(user_query, context, max_new_tokens=300):
    """
    Generate a recipe suggestion or customisation based on user input.

    Args:
    - user_query: User's customisation request or question.
    - context: Contextual information (e.g., recipe details).
    - max_new_tokens: Number of new tokens to generate.

    Returns:
    - Generated response from the LLM.
    """
    # Combine the user query with the recipe context
    input_text = f"Context:\n{context}\n\nUser Query:\n{user_query}\n\nResponse:"

    # Tokenize input with padding and truncation
    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512  # Adjust max length for tokenized input
    )

    # Generate response with sampling enabled
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],  # Add attention mask
        max_new_tokens=max_new_tokens,  # Specify new token limit
        pad_token_id=tokenizer.eos_token_id,  # Set pad token to eos token
        temperature=0.7,  # Reduce randomness
        top_k=50,  # Focus on top 50 tokens
        top_p=0.9,  # Nucleus sampling
        do_sample=True  # Enable sampling for creative outputs
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return response

# Example: User selects a recipe
recipe_index = 0  # Select the first recipe
ingredients = df_cleaned['ingredients'][recipe_index][:100]  # Truncate ingredients
directions = df_cleaned['directions'][recipe_index][:100]  # Truncate directions

recipe_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                 f"Ingredients: {ingredients}\n" \
                 f"Directions: {directions}"

# Example user query
user_query = "Make this recipe vegan and suggest a side dish."

# Get the customised recipe suggestion
customised_response = get_recipe_customisation(user_query, recipe_context, max_new_tokens=300)

print("Customised Recipe Response:")
print(customised_response)



Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Model and tokenizer loaded successfully!
Customised Recipe Response:
Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, c
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine 

User Query:
Make this recipe vegan and suggest a side dish.

Response:
<pre><code>Vegan Gazpacho Recipe
Ingredients:
- 8 tomatoes, quartered
- 2 onions, diced
- 1 bell pepper, diced
- 2 cloves of garlic, minced
- 1/2 cup olive oil
- 1/2 cup red wine vinegar
- 1 teaspoon salt
- 1 teaspoon sugar
- 1 teaspoon paprika
- 1/4 cup chopped fresh parsley
- 2 avocados, diced

Directions:
1. In a food processor, combine the tomatoes, onions, bell pepper, garlic, olive oil, red wine vinegar, salt, sugar, and paprika. Process until smooth.
2. Transfer the mixture to a large bowl and stir in the parsley.
3. In a separate bowl, mash the avocados with a fork and add to the tomato mixtu

What’s Working Well
Creative Adjustments:
The recipe includes thoughtful substitutions and additions, such as avocados for creaminess.
Concise Yet Informative:
Both the veganisation process and side dish suggestion are included.


Extend the Functionality
Dynamic Queries:

Allow users to ask varied questions like:
"Reduce the cooking time."
"Add garlic to this recipe."
"Suggest alternative ingredients."

What to Add for Dynamic Queries
Dynamic queries allow users to interact with the system beyond simple recipe suggestions. Below are potential features and the business needs they solve:

1. Multi-step Query Handling
Why? Users might want to ask follow-up questions or refine the recipe further.

In [ ]:
# Initialize the session context with the selected recipe
session_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                  f"Ingredients: {df_cleaned['ingredients'][recipe_index][:100]}\n" \
                  f"Directions: {df_cleaned['directions'][recipe_index][:100]}"

# Function to handle dynamic queries
def handle_follow_up_query(user_query, session_context, max_new_tokens=200):
    """
    Process a follow-up query dynamically by appending it to the session context.
    """
    input_text = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:"

    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Update the session context with the user's query and model's response
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

# Example: User follows up with another query
user_query = "Suggest a wine pairing for this recipe."
customised_response, session_context = handle_follow_up_query(user_query, session_context)
print("Customised Response:", customised_response)


Customised Response: Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, c
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine 

User Query:
Suggest a wine pairing for this recipe.

Response:
A light red wine, such as a Pinot Noir or a Beaujolais, would pair nicely with this recipe.


Dietary and Nutritional Filters Why? Customisation based on user preferences or dietary needs is a key business goal in food tech.

In [ ]:
# Initialize the session context with the selected recipe
session_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                  f"Ingredients: {df_cleaned['ingredients'][recipe_index][:100]}\n" \
                  f"Directions: {df_cleaned['directions'][recipe_index][:100]}"

# Function to handle dynamic queries
def handle_follow_up_query(user_query, session_context, max_new_tokens=200):
    """
    Process a follow-up query dynamically by appending it to the session context.
    """
    input_text = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:"

    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Update the session context with the user's query and model's response
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

# Example dietary query
dietary_query = "Make this recipe gluten-free."

# Generate response for dietary customisation
customised_dietary_response, session_context = handle_follow_up_query(dietary_query, session_context)
print("Dietary Customisation Response:", customised_dietary_response)

Dietary Customisation Response: Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, c
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine 

User Query:
Make this recipe gluten-free.

Response:
<pre><code>['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth", "Combine 

Gazpacho with 8 tomatoes, quartered, 1 red onion, cut into small dice, and 1 green bell pepper, chopped. Mix with 1/4 cup olive oil, 1/4 cup tomato juice, 1/4 cup red wine vinegar, 1/4 cup honey, 1 teaspoon fresh thyme, 1/2 teaspoon kosher salt, and 1/2 teaspoon freshly ground black pepper. Puree until smooth.

Add the tomatoes to a food processor with a pinch of salt and puree until smooth. Combine with the rest of the ingredients and process until smooth.

Make this recipe gluten-free by using cornstarch instead of wheat flour to thicken the sou

Ingredient Substitutions
Why? Allow users to modify recipes dynamically based on available ingredients or preferences.

In [ ]:
# Initialize the session context with the selected recipe
session_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                  f"Ingredients: {df_cleaned['ingredients'][recipe_index][:100]}\n" \
                  f"Directions: {df_cleaned['directions'][recipe_index][:100]}"

# Function to handle dynamic queries
def handle_follow_up_query(user_query, session_context, max_new_tokens=200):
    """
    Process a follow-up query dynamically by appending it to the session context.
    """
    input_text = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:"

    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Update the session context with the user's query and model's response
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

# Example dietary query
dietary_query = "Make this recipe gluten-free."

ingredient_query = "Replace olive oil with coconut oil."

# Generate response for ingredient substitution
ingredient_substitution_response, session_context = handle_follow_up_query(ingredient_query, session_context)
print("Ingredient Substitution Response:", ingredient_substitution_response)


Ingredient Substitution Response: Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, c
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine 

User Query:
Replace olive oil with coconut oil.

Response:
<pre><code>['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth", "Combine

Gazpacho
User 


Multi-recipe Comparisons
Why? Let users compare two recipes and choose the one that fits their needs better.



In [ ]:
# Initialize the session context with the selected recipe
session_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                  f"Ingredients: {df_cleaned['ingredients'][recipe_index][:100]}\n" \
                  f"Directions: {df_cleaned['directions'][recipe_index][:100]}"

# Function to handle dynamic queries
def handle_follow_up_query(user_query, session_context, max_new_tokens=200):
    """
    Process a follow-up query dynamically by appending it to the session context.
    """
    input_text = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:"

    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Update the session context with the user's query and model's response
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

# Example: Compare two recipes
recipe_index_1 = 0
recipe_index_2 = 1

recipe_context_1 = f"Title: {df_cleaned['title'][recipe_index_1]}\nIngredients: {df_cleaned['ingredients'][recipe_index_1]}"
recipe_context_2 = f"Title: {df_cleaned['title'][recipe_index_2]}\nIngredients: {df_cleaned['ingredients'][recipe_index_2]}"

comparison_query = "Which recipe is quicker to prepare?"
comparison_context = f"Recipe 1:\n{recipe_context_1}\n\nRecipe 2:\n{recipe_context_2}"
comparison_response, _ = handle_follow_up_query(comparison_query, comparison_context)
print("Comparison Response:", comparison_response)


Comparison Response: Context:
Recipe 1:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, cut into small dice', '1 red bell pepper, cut into small dice', '1 yellow bell pepper, cut into small dice', '1/2 cucumber, cut into small dice', 'Extra-virgin olive oil, for drizzling', '3 leaves fresh basil, finely chopped']

Recipe 2:
Title: Rice Paper Wrapped Halibut Roasted with Foie Gras Gingered Spaghetti Squash and Wild Mushroom Ragout
Ingredients: ['4 large, round rice wrappers', '4 (6-ounce) fillets of halibut', '8 ounces foie gras mousse', '12 long chive batons', 'Salt and black pepper', 'Chives, for garnishing', 'Truffle oil for garnishing', '8 ounces B or C grade fresh foie gras', '1 whole egg', '1/4 cup heavy cream', '1 teaspoon white truffle oil', '4 ounces diced chicken breast, skinless', '1 tablespoon chives, finely chopped', 'Salt and black pepper', '1 small spaghetti squash cut in half, deseeded', '2 t

Dietary Restriction Recommendations
Why? Address diverse user needs by suggesting alternative recipes for specific diets.

In [ ]:
# Initialize the session context with the selected recipe
session_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                  f"Ingredients: {df_cleaned['ingredients'][recipe_index][:100]}\n" \
                  f"Directions: {df_cleaned['directions'][recipe_index][:100]}"

# Function to handle dynamic queries
def handle_follow_up_query(user_query, session_context, max_new_tokens=200):
    """
    Process a follow-up query dynamically by appending it to the session context.
    """
    input_text = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:"

    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Update the session context with the user's query and model's response
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

dietary_recommendation_query = "Suggest three vegan recipes for dinner."

# Generate recommendations
vegan_recommendations, session_context = handle_follow_up_query(dietary_recommendation_query, session_context)
print("Vegan Recommendations:", vegan_recommendations)

Vegan Recommendations: Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, c
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine 

User Query:
Suggest three vegan recipes for dinner.

Response:
1. Vegan lentil soup with roasted vegetables: 
- Ingredients: ['1 cup lentils', '2 tablespoons olive oil', '1 onion, diced', '2 carrots, diced', '2 bell peppers, diced', '1 cup mushrooms', '1 garlic clove, minced', '2 tablespoons tomato paste', '2 cups vegetable stock', '2 tablespoons nutritional yeast', '1 teaspoon thyme', '1 teaspoon oregano', '1 teaspoon oregano', '1 teaspoon salt', '1 teaspoon pepper', '1 teaspoon paprika', '2 tablespoons olive oil', '1 onion, diced', '2 carrots, diced', '2 bell peppers, diced', '1 garlic clove, minced', '2 tablespoons tomato paste', '2 cups vegetable stock', '2 tablespoons nutritional yeast', '1 teaspoon thyme', '1

In [ ]:
# Initialize the session context with the selected recipe
recipe_index = 0  # Example recipe index
session_context = f"Title: {df_cleaned['title'][recipe_index]}\n" \
                  f"Ingredients: {str(df_cleaned['ingredients'][recipe_index])[:100]}\n" \
                  f"Directions: {str(df_cleaned['directions'][recipe_index])[:100]}"

# Function to handle dynamic queries
def handle_follow_up_query(user_query, session_context, max_new_tokens=200):
    """
    Process a follow-up query dynamically by appending it to the session context.
    """
    # Combine session context and user query
    input_text = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:"

    # Tokenize input with truncation and padding
    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    # Generate response with repetition penalty and adjusted sampling
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.8,  # Slightly increase randomness
        top_k=50,  # Focus on top 50 tokens
        top_p=0.9,  # Nucleus sampling
        repetition_penalty=1.2,  # Penalise repetitive tokens
        do_sample=True
    )

    # Decode and update session context
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

# Example query for dietary recommendations
dietary_recommendation_query = "Suggest three vegan recipes for dinner."

# Generate recommendations
vegan_recommendations, session_context = handle_follow_up_query(dietary_recommendation_query, session_context)

# Display the response
print("Vegan Recommendations:")
print(vegan_recommendations)


Vegan Recommendations:
Context:
Title: Gazpacho
Ingredients: ['8 tomatoes, quartered', 'Kosher salt', '1 red onion, cut into small dice', '1 green bell pepper, c
Directions: ['["Add the tomatoes to a food processor with a pinch of salt and puree until smooth', '", "Combine 

User Query:
Suggest three vegan recipes for dinner.

Response:
1. Vegan Chili
2. Vegan Lentil Soup
3. Vegan Tofu Stir Fry


In [ ]:
# Function to filter recipes based on a keyword
def filter_recipes_by_keyword(keyword, dataset, max_results=3):
    """
    Filter recipes in the dataset based on a keyword (e.g., 'vegan').

    Args:
    - keyword: The keyword to filter recipes (e.g., 'vegan').
    - dataset: The dataset containing recipes.
    - max_results: Maximum number of recipes to return.

    Returns:
    - A list of filtered recipes with title, ingredients, and directions.
    """
    filtered_recipes = dataset[dataset['ingredients'].str.contains(keyword, case=False, na=False)]
    return filtered_recipes.head(max_results)

# Function to dynamically respond with recipes
def handle_query_with_recipes(user_query, session_context, keyword, dataset, max_results=3):
    """
    Process a query and dynamically return recipes that match the keyword.

    Args:
    - user_query: User's query string.
    - session_context: Current session context.
    - keyword: Keyword to filter recipes.
    - dataset: Recipe dataset.
    - max_results: Maximum number of recipes to return.

    Returns:
    - Generated response with matching recipes and updated session context.
    """
    # Filter recipes based on keyword
    matching_recipes = filter_recipes_by_keyword(keyword, dataset, max_results)

    # Format the response with recipe details
    if not matching_recipes.empty:
        response = "Here are some recipes:\n"
        for idx, row in matching_recipes.iterrows():
            response += f"Title: {row['title']}\n"
            response += f"Ingredients: {row['ingredients']}\n"
            response += f"Directions: {row['directions']}\n\n"
    else:
        response = f"Sorry, no recipes found for '{keyword}'."

    # Update session context
    session_context += f"\n\nUser Query: {user_query}\nResponse: {response}"
    return response, session_context

# Example: User asks for vegan recipes
dietary_recommendation_query = "Suggest three vegan recipes for dinner."
keyword = "vegan"  # Filter for 'vegan' recipes
max_results = 3

# Generate recommendations
vegan_recommendations, session_context = handle_query_with_recipes(
    dietary_recommendation_query, session_context, keyword, df_cleaned, max_results
)

# Display the response
print("Vegan Recommendations:")
print(vegan_recommendations)


Vegan Recommendations:
Here are some recipes:
Title: Vegan Chocolate Ganache Cupcakes with Salted Caramel and Dark Chocolate Buttercream
Ingredients: ['1 1/2 cups soy milk, or non-dairy milk of your choice', '3/4 cup agave nectar or maple syrup', '1/4 cup plus 2 tablespoons canola oil', '1/4 cup plus 2 tablespoons organic cane sugar', '1 1/2 teaspoons vanilla extract', '1 1/2 teaspoons apple cider vinegar', '1 1/2 cups organic, unbleached white flour', '1/4 cup plus 2 tablespoons cocoa powder', '1/4 cup dark cocoa powder', '1 1/2 teaspoons baking powder', '3/4 teaspoon baking soda', '1/4 teaspoon salt', '4 cups dark chocolate chunks', '1 cup soy milk, or non-dairy milk of your choice', '1/2 cup agave nectar or maple syrup', '1 1/2 cups coconut milk fat (from two 13.5-ounce cans unsweetened, full fat coconut milk that has been refrigerated overnight)', '1 cup organic brown sugar', '1 tablespoon vanilla extract', '2 teaspoons course sea salt', '2/3 cup organic vegetable shortening', '1/3

In [ ]:
# Function to filter recipes by a single keyword
def filter_recipes_by_keyword(keyword, dataset, max_results=3):
    """
    Filter recipes in the dataset based on a single keyword (e.g., 'vegan').

    Args:
    - keyword: The keyword to filter recipes (e.g., 'vegan').
    - dataset: The dataset containing recipes.
    - max_results: Maximum number of recipes to return.

    Returns:
    - A filtered DataFrame of recipes.
    """
    filtered_recipes = dataset[dataset['ingredients'].str.contains(keyword, case=False, na=False)]
    return filtered_recipes.head(max_results)


# Function to filter recipes by multiple keywords
def filter_recipes_by_multiple_keywords(keywords, dataset, max_results=3):
    """
    Filter recipes in the dataset based on multiple keywords (e.g., 'vegan', 'gluten-free').

    Args:
    - keywords: A list of keywords to filter recipes (e.g., ['vegan', 'gluten-free']).
    - dataset: The dataset containing recipes.
    - max_results: Maximum number of recipes to return.

    Returns:
    - A filtered DataFrame of recipes.
    """
    filtered_recipes = dataset
    for keyword in keywords:
        filtered_recipes = filtered_recipes[filtered_recipes['ingredients'].str.contains(keyword, case=False, na=False)]
    return filtered_recipes.head(max_results)


# Function to dynamically handle dietary recommendation queries
def handle_query_with_recipes(user_query, dataset, max_results=3):
    """
    Process a user query and dynamically return recipes based on keywords.

    Args:
    - user_query: The user query string.
    - dataset: The dataset containing recipes.
    - max_results: Maximum number of recipes to return.

    Returns:
    - A user-friendly string response with recipe recommendations.
    """
    # Example: Extract keywords from user query (for simplicity, we'll hardcode the keywords here)
    if "vegan and gluten-free" in user_query.lower():
        keywords = ["vegan", "gluten-free"]
    elif "vegan" in user_query.lower():
        keywords = ["vegan"]
    elif "gluten-free" in user_query.lower():
        keywords = ["gluten-free"]
    else:
        keywords = []  # No specific keywords detected

    # Filter recipes based on keywords
    if keywords:
        filtered_recipes = filter_recipes_by_multiple_keywords(keywords, dataset, max_results)
    else:
        return "Sorry, I couldn't detect any dietary preferences in your query."

    # Format the response
    if not filtered_recipes.empty:
        response = "Here are some recipes based on your preferences:\n\n"
        for idx, row in filtered_recipes.iterrows():
            response += f"Title: {row['title']}\n"
            response += f"Ingredients: {row['ingredients']}\n"
            response += f"Directions: {row['directions']}\n\n"
    else:
        response = f"Sorry, no recipes found for keywords: {', '.join(keywords)}."

    return response


# Example Dataset and Query
# Assume `df_cleaned` is your pre-loaded recipe dataset
dietary_recommendation_query = "Suggest three vegan and gluten-free recipes for dinner."
max_results = 3

# Generate recommendations dynamically
recommendations = handle_query_with_recipes(dietary_recommendation_query, df_cleaned, max_results)

# Display the final response
print("Response to User Query:")
print(recommendations)


Response to User Query:
Here are some recipes based on your preferences:

Title: Chipotle Dark Chocolate Cupcakes
Ingredients: ['1 cup gluten-free flour mix', '3/4 cup granulated sugar', '1/2 cup cocoa powder', '1 teaspoon baking powder', '1/2 teaspoon chipotle powder', '1/2 teaspoon ground cinnamon', '1/2 teaspoon sea salt', '1/2 teaspoon xanthan gum', '1 1/4 cups soy milk', '1/2 cup canola oil', '1/4 cup espresso', '1 tablespoon vanilla extract', '1 tablespoon vinegar', '1 chipotle dark chocolate bar, cut into 12 large chunks', '8 cups powdered sugar', '1 cup cocoa powder', '8 ounces vegan butter spread, such as Earth Balance', '3/4 cup soy milk', '1 tablespoon vanilla extract']
Directions: ['["For the cupcakes: Preheat the oven to 350 degrees F', ' Line a standard cupcake or muffin pan with 12 cupcake liners', '", "In medium bowl, whisk together with an electric mixer or hand whisk the flour mix, granulated sugar, cocoa, baking powder, chipotle powder, cinnamon, salt and xanthan gum

In [ ]:
# Example: Categorising recipes based on title or directions
df_cleaned['is_baking'] = df_cleaned['title'].str.contains("cupcake|cake|cookie|bake|brownie", case=False, na=False) | \
                          df_cleaned['directions'].str.contains("bake|oven", case=False, na=False)

# Count baking vs non-baking recipes
baking_count = df_cleaned['is_baking'].sum()
non_baking_count = len(df_cleaned) - baking_count

print(f"Baking recipes: {baking_count}")
print(f"Non-baking recipes: {non_baking_count}")


Baking recipes: 20559
Non-baking recipes: 28882


In [ ]:
# Exclude recipes containing baking keywords
non_baking_recipes = df_cleaned[~df_cleaned['directions'].str.contains("bake|oven", case=False, na=False)]
filtered_recipes = non_baking_recipes[
    non_baking_recipes['ingredients'].str.contains("vegan", case=False, na=False) &
    non_baking_recipes['ingredients'].str.contains("gluten-free", case=False, na=False)
]


In [ ]:
def generate_non_baking_recipes(keywords, max_results=3):
    prompt = f"Suggest {max_results} vegan and gluten-free recipes that are not baking. Provide title, ingredients, and directions."
    inputs = tokenizer(
        prompt,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=300,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return response

# Example
synthetic_recipes = generate_non_baking_recipes(["vegan", "gluten-free"], max_results=3)
print("Generated Non-Baking Recipes:", synthetic_recipes)


Generated Non-Baking Recipes: Suggest 3 vegan and gluten-free recipes that are not baking. Provide title, ingredients, and directions.
1. Vegan Quinoa Salad: Cook 1 cup of quinoa according to package instructions, then mix with diced cucumber, cherry tomatoes, red bell pepper, red onion, and a handful of fresh basil. Toss with a zesty lemon vinaigrette made from olive oil, lemon juice, Dijon mustard, garlic, salt, and pepper. Directions: In a small bowl, whisk together all the dressing ingredients. In a separate bowl, combine the quinoa and vegetable mixture. Gently toss the quinoa mixture in the dressing to coat, then serve.

2. Sweet Potato and Black Bean Chili: Roast 2 sweet potatoes, cut into small cubes, in a 400°F oven for 20 minutes. Meanwhile, cook 2 cups of black beans according to package instructions. Drain the beans and add them to the sweet potato cubes. In a small bowl, combine the sweet potato and bean mixture with a pinch of cumin, paprika, and chili powder. Gently toss

Next Steps
Test the LLM’s outputs for diverse queries and ensure they align with user expectations.
Deploy the system via a web interface (e.g., using Flask or Streamlit).
Explore adding more advanced capabilities like dietary filtering or multilingual support.

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the Falcon-7B-Instruct model
model_name = "tiiuae/falcon-7b-instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Add padding token
tokenizer.pad_token = tokenizer.eos_token

# Function to query the LLM
def query_llm(prompt, max_new_tokens=150):
    """
    Generate a response from the LLM based on the input prompt.

    Args:
    - prompt: The text prompt for the LLM.
    - max_new_tokens: Number of new tokens to generate.

    Returns:
    - The generated response from the LLM.
    """
    inputs = tokenizer(
        prompt,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return response

# Test diverse queries
def test_llm_responses(test_cases, session_context=None):
    """
    Test the LLM with diverse queries.

    Args:
    - test_cases: List of test cases with prompts and expected outcomes.
    - session_context: Optional session context to simulate a dialogue.

    Returns:
    - Dictionary of results for manual review.
    """
    results = {}
    for i, test_case in enumerate(test_cases, start=1):
        user_query = test_case.get("query")
        prompt = f"Context:\n{session_context}\n\nUser Query:\n{user_query}\n\nResponse:" if session_context else user_query
        print(f"Testing Query {i}: {user_query}")
        response = query_llm(prompt)
        results[f"Test Case {i}"] = {
            "query": user_query,
            "response": response,
            "expected_behavior": test_case.get("expected_behavior", "Not Provided")
        }
    return results

# Example diverse test cases
test_cases = [
    {
        "query": "Suggest three recipes for a gluten-free and low-carb dinner.",
        "expected_behavior": "LLM provides three valid recipes with ingredients and cooking methods."
    },
    {
        "query": "How can I make Gazpacho spicier?",
        "expected_behavior": "LLM suggests ways to make Gazpacho spicier, such as adding chili or hot sauce."
    },
    {
        "query": "Recommend a vegan dessert with chocolate.",
        "expected_behavior": "LLM provides a vegan chocolate dessert recipe with clear directions."
    },
    {
        "query": "List quick snacks for a picnic.",
        "expected_behavior": "LLM provides a list of quick snack ideas suitable for a picnic."
    },
    {
        "query": "What wine pairs well with mushroom risotto?",
        "expected_behavior": "LLM suggests appropriate wine pairings like Pinot Noir or Chardonnay."
    }
]

# Optional session context
session_context = "Title: Mushroom Risotto\nIngredients: ['Arborio rice', 'Mushrooms', 'Vegetable stock', 'Parmesan', 'Butter', 'Garlic']\nDirections: ['Sauté garlic and mushrooms in butter. Add rice and stock gradually while stirring. Finish with Parmesan.']"

# Run tests
results = test_llm_responses(test_cases, session_context)

# Display results
for test_case, result in results.items():
    print(f"\n{test_case}:")
    print(f"Query: {result['query']}")
    print(f"Response: {result['response']}")
    print(f"Expected Behavior: {result['expected_behavior']}")


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Testing Query 1: Suggest three recipes for a gluten-free and low-carb dinner.
Testing Query 2: How can I make Gazpacho spicier?
Testing Query 3: Recommend a vegan dessert with chocolate.
Testing Query 4: List quick snacks for a picnic.
Testing Query 5: What wine pairs well with mushroom risotto?

Test Case 1:
Query: Suggest three recipes for a gluten-free and low-carb dinner.
Response: Context:
Title: Mushroom Risotto
Ingredients: ['Arborio rice', 'Mushrooms', 'Vegetable stock', 'Parmesan', 'Butter', 'Garlic']
Directions: ['Sauté garlic and mushrooms in butter. Add rice and stock gradually while stirring. Finish with Parmesan.']

User Query:
Suggest three recipes for a gluten-free and low-carb dinner.

Response:
1. Grilled salmon with roasted vegetables
2. Baked chicken with sweet potato and green beans
3. Low-carb spaghetti with homemade meatballs and marinara sauce
Expected Behavior: LLM provides three valid recipes with ingredients and cooking methods.

Test Case 2:
Query: How can I

In [ ]:
def get_recipe_customisation(user_query, context, max_new_tokens=200):
    """
    Generate a recipe suggestion or customisation based on user input.

    Args:
    - user_query: User's customisation request or question.
    - context: Contextual information (e.g., recipe details).
    - max_new_tokens: Number of new tokens to generate.

    Returns:
    - Generated response from the LLM.
    """
    if not context.strip():
        context = "General cooking context: Please ensure your query is specific for better suggestions."

    # Combine the user query with the recipe context
    input_text = f"Context:\n{context}\n\nUser Query:\n{user_query}\n\nResponse:"

    # Tokenize input
    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )

    # Generate response
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )

    response = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Append the user query and response to the context for continuity
    context += f"\n\nUser Query: {user_query}\nResponse: {response}"

    return response, context


In [ ]:
def recommend_recipes(user_ingredients, embeddings, df_cleaned, top_n=5):
    """
    Recommend recipes based on user ingredients.

    Args:
    - user_ingredients: String containing user ingredients (e.g., "chicken, garlic, onion").
    - embeddings: Precomputed embeddings for recipes.
    - df_cleaned: DataFrame containing recipe details.
    - top_n: Number of recommendations to return.

    Returns:
    - DataFrame with recommended recipes and their similarity scores.
    """
    # Generate query embedding for user ingredients
    query_embedding = get_query_embedding(user_ingredients)

    # Calculate cosine similarity between query and all recipe embeddings
    similarities = cosine_similarity([query_embedding], embeddings)[0]

    # Get indices of top N most similar recipes
    top_indices = similarities.argsort()[-top_n:][::-1]  # Sort and reverse

    # Extract recommended recipes
    recommended_recipes = df_cleaned.iloc[top_indices].copy()
    recommended_recipes['similarity_score'] = similarities[top_indices]

    # Add detailed directions and tips to enhance user experience
    for idx in recommended_recipes.index:
        recipe = recommended_recipes.loc[idx]
        recipe_title = recipe['title']
        recipe_directions = recipe['directions']
        recommended_recipes.loc[idx, 'detailed_response'] = (
            f"Title: {recipe_title}\n"
            f"Directions: {recipe_directions}\n"
            f"Serving Tips: This dish can be paired with a light salad or bread for a wholesome meal."
        )

    return recommended_recipes[['title', 'ingredients', 'directions', 'detailed_response', 'similarity_score']]


In [ ]:
def recommend_diverse_recipes(user_ingredients, embeddings, df_cleaned, top_n=5, diversity_field='type'):
    """
    Recommend diverse recipes based on user ingredients and a diversity field.

    Args:
    - user_ingredients: User-provided ingredients (e.g., "mushrooms, garlic").
    - embeddings: Precomputed embeddings for recipes.
    - df_cleaned: DataFrame with recipe details.
    - top_n: Number of recommendations to return.
    - diversity_field: Field to ensure diversity (e.g., "type" for baking vs. non-baking).

    Returns:
    - DataFrame with diverse recipe recommendations.
    """
    # Recommend recipes based on embeddings
    recommended_recipes = recommend_recipes(user_ingredients, embeddings, df_cleaned, top_n * 2)

    # Group and sort by diversity field
    grouped = recommended_recipes.groupby(diversity_field).apply(lambda x: x.head(top_n // 2))
    grouped = grouped.sample(frac=1).reset_index(drop=True)

    # Ensure unique and detailed responses
    grouped['detailed_response'] = grouped.apply(
        lambda x: (
            f"Title: {x['title']}\n"
            f"Ingredients: {x['ingredients']}\n"
            f"Directions: {x['directions']}\n"
            f"Tips: For a creative twist, try adding complementary herbs or sauces!"
        ), axis=1
    )

    return grouped.head(top_n)


In [ ]:
def test_llm_responses(queries, session_context, max_results=3):
    """
    Test LLM responses for diverse queries with simplified output.

    Args:
    - queries: List of user queries.
    - session_context: Current session context.
    - max_results: Number of results to generate for each query.

    Returns:
    - Dictionary of queries and their respective responses.
    """
    results = {}
    for query in queries:
        try:
            response, session_context = get_recipe_customisation(query, session_context, max_new_tokens=300)
            results[query] = response
        except Exception as e:
            results[query] = f"Error: {str(e)}"
    return results


# Reduced Example Queries
queries = [
    "Suggest three recipes for a gluten-free and low-carb dinner.",
    "How can I make Gazpacho spicier?",
    "Recommend a vegan dessert with chocolate."
]

# Test LLM Responses
session_context = "Context: Recipe recommendations and customisations."
responses = test_llm_responses(queries, session_context)

# Display Results in a Clear Format
for query, response in responses.items():
    print(f"Query: {query}\nResponse:\n{response}\n{'-'*40}")


Query: Suggest three recipes for a gluten-free and low-carb dinner.
Response:
Context:
Context: Recipe recommendations and customisations.

User Query:
Suggest three recipes for a gluten-free and low-carb dinner.

Response:
1. Baked salmon with avocado salsa. 
2. Cauliflower and chickpea curry with quinoa. 
3. Zucchini noodles with tomato sauce and meatballs.
----------------------------------------
Query: How can I make Gazpacho spicier?
Response:
Context:
Context: Recipe recommendations and customisations.

User Query: Suggest three recipes for a gluten-free and low-carb dinner.
Response: Context:
Context: Recipe recommendations and customisations.

User Query:
Suggest three recipes for a gluten-free and low-carb dinner.

Response:
1. Baked salmon with avocado salsa. 
2. Cauliflower and chickpea curry with quinoa. 
3. Zucchini noodles with tomato sauce and meatballs.

User Query:
How can I make Gazpacho spicier?

Response:
To make Gazpacho spicier, you can add diced jalapenos or red 

Final Implementation
Features:
 - Batch Processing:

Combines queries into a single request for efficiency.
Saves API costs and reduces processing time.

 - Caching:

Stores results for repeated queries.
Prevents redundant computation.

 - Asynchronous Execution:

Handles multiple queries concurrently.
Improves performance for large batches.

 - Flexible Execution:

Automatically chooses the best method based on query size and performance needs.

Achievements:
Performance Optimisation:

Batch Processing: Processes multiple queries in a single API call, reducing latency and network overhead.
Caching: Uses lru_cache to store previously computed results, which speeds up repeated queries and saves API resources.
Asynchronous Processing: Allows concurrent query handling, enabling quicker responses for multiple queries.
Real-Time Execution:

Using asynchronous execution (asyncio and nest_asyncio), the system can handle multiple queries in real-time without blocking operations.
Flexibility:

Adapts to the size and complexity of the input queries.
Allows toggling between batch processing and async processing based on the use_async parameter.
Scalability:

The solution is designed to handle multiple user queries efficiently, whether they are processed sequentially, in batches, or asynchronously.
Ease of Debugging:

The modular nature of the system (batch_process_queries, get_cached_response, and async_query) makes it easy to isolate and troubleshoot components if issues arise.

In [ ]:
import asyncio
from functools import lru_cache
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load the Falcon-7B-Instruct model from Hugging Face
model_name = "tiiuae/falcon-7b-instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Add padding token to the tokenizer
tokenizer.pad_token = tokenizer.eos_token
print("Model and tokenizer loaded successfully!")


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Model and tokenizer loaded successfully!


In [ ]:
def get_recipe_customisation(user_query, session_context, max_new_tokens=300):
    """
    Generate a recipe suggestion or customisation based on user input.
    """
    # Combine the query with the session context
    input_text = f"Context: {session_context}\n\nUser Query: {user_query}\n\nResponse:"
    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=512
    )
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )
    return tokenizer.decode(outputs[0], skip_special_tokens=True)


In [ ]:
def batch_process_queries(queries, session_context, max_new_tokens=300):
    """
    Process multiple queries in a single API call to optimise performance.
    """
    combined_prompt = "\n\n".join([f"Query {i+1}: {query}" for i, query in enumerate(queries)])
    input_text = f"Context: {session_context}\n\nBatch Queries:\n{combined_prompt}\n\nResponses:"

    inputs = tokenizer(
        input_text,
        return_tensors="pt",
        truncation=True,
        padding=True,
        max_length=1024
    )
    outputs = model.generate(
        inputs["input_ids"],
        attention_mask=inputs["attention_mask"],
        max_new_tokens=max_new_tokens,
        pad_token_id=tokenizer.eos_token_id,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        do_sample=True
    )
    response_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    response_lines = response_text.split("\n")
    return {queries[i]: response_lines[i + 1] for i in range(len(queries))}


In [ ]:
@lru_cache(maxsize=128)
def get_cached_response(query, session_context, max_new_tokens=300):
    """
    Cached version of the recipe customisation function.
    """
    return get_recipe_customisation(query, session_context, max_new_tokens)


In [ ]:
!pip install nest_asyncio

In [ ]:
import nest_asyncio
nest_asyncio.apply()


In [ ]:
async def async_query(query, session_context, max_new_tokens=300):
    """
    Perform an asynchronous query to the LLM.
    """
    return get_recipe_customisation(query, session_context, max_new_tokens)

async def process_queries_async(queries, session_context, max_new_tokens=300):
    """
    Process multiple queries asynchronously.
    """
    tasks = [async_query(query, session_context, max_new_tokens) for query in queries]
    results = await asyncio.gather(*tasks)
    return {queries[i]: results[i] for i in range(len(queries))}


In [ ]:
def combined_test_llm_responses(queries, session_context, use_async=True, max_new_tokens=300):
    """
    Test LLM responses with batch processing, caching, and optional async execution.
    """
    if use_async:
        return asyncio.run(process_queries_async(queries, session_context, max_new_tokens))
    elif len(queries) > 1:
        return batch_process_queries(queries, session_context, max_new_tokens)
    else:
        return {query: get_cached_response(query, session_context, max_new_tokens) for query in queries}


In [ ]:
import asyncio

# Queries to test
queries = [
    "Suggest three recipes for a gluten-free and low-carb dinner.",
    "How can I make Gazpacho spicier?",
    "Recommend a vegan dessert with chocolate."
]

# Context for the session
session_context = "Context: Recipe recommendations and customisations."

# Ensure the function handles async internally
responses = combined_test_llm_responses(queries, session_context, use_async=True)

# Display responses
for query, response in responses.items():
    print(f"Query: {query}\nResponse: {response}\n{'-'*40}")




Query: Suggest three recipes for a gluten-free and low-carb dinner.
Response: Context: Context: Recipe recommendations and customisations.

User Query: Suggest three recipes for a gluten-free and low-carb dinner.

Response:
1. Grilled salmon with roasted vegetables: This recipe is quick and easy to prepare, and is the perfect low-carb meal. The salmon is grilled and served with roasted vegetables such as broccoli, asparagus, and bell peppers, drizzled with olive oil and lemon juice.

2. Eggplant parmesan: This classic Italian dish is a great low-carb option. Eggplant is coated in breadcrumbs and Parmesan cheese and baked until crispy. It can be served with a side of marinara sauce and melted mozzarella cheese.

3. Low-carb chicken fajitas: Chicken, bell peppers, and onions are cooked in a low-carb friendly skillet with spices like cumin and paprika. The dish is served with a side of rice, making it a complete meal.
----------------------------------------
Query: How can I make Gazpacho

Backend API Development and Frontend  or API Demonstration in Postman

Performance monitoring: Select metrics that quantify the quality and relevance of the LLM outputs.

Relevance Score: Measure semantic similarity between LLM responses and expected responses using cosine similarity with embeddings.
Execution Time: Monitor the time taken to generate responses.
Response Diversity: Evaluate how varied the responses are for similar queries (e.g., by comparing Jaccard similarity or using n-grams).
Response Completeness: Check if all parts of a response are present (e.g., ingredients, directions, and tips in recipe responses).

Implement Automated Evaluation with Metrics
Create a function that evaluates LLM outputs based on the defined metrics:

In [ ]:
import numpy as np
from sklearn.metrics import accuracy_score

def evaluate_llm_responses(queries, responses, expected_outputs, metrics):
    """
    Evaluate LLM responses based on defined metrics.

    Args:
    - queries: List of user queries.
    - responses: List of responses from the LLM.
    - expected_outputs: Expected outputs or ground truth for comparison.
    - metrics: Dictionary defining evaluation criteria.

    Returns:
    - A dictionary with evaluation scores for each metric.
    """
    evaluation_results = {}

    # Relevance: Check if key phrases from the query are in the response
    def relevance_score(query, response):
        return len(set(query.lower().split()) & set(response.lower().split())) / len(query.split())

    # Completeness: Check if the response includes required details
    def completeness_score(response, expected):
        return len(set(expected.lower().split()) & set(response.lower().split())) / len(expected.split())

    # Diversity: Check response uniqueness by comparing responses
    def diversity_score(responses):
        unique_responses = set(responses)
        return len(unique_responses) / len(responses)

    # Accuracy: Compare response with expected output
    def accuracy_score(response, expected):
        return 1 if expected in response else 0

    # Evaluate each metric
    for query, response, expected in zip(queries, responses, expected_outputs):
        evaluation_results[query] = {
            "relevance": relevance_score(query, response),
            "completeness": completeness_score(response, expected),
            "accuracy": accuracy_score(response, expected)
        }

    # Add diversity as a global metric
    evaluation_results["diversity"] = diversity_score(responses)

    return evaluation_results



Log and Visualise Performance Over Time
Implement a logging mechanism to track performance trends:

In [ ]:
import matplotlib.pyplot as plt

def log_and_visualise_performance(metrics_over_time):
    """
    Visualise performance metrics over time to monitor trends.

    Args:
    - metrics_over_time: A dictionary with timestamps and evaluation results.

    Returns:
    - Performance trend visualisation.
    """
    timestamps = list(metrics_over_time.keys())
    relevance_scores = [metrics["relevance"] for metrics in metrics_over_time.values()]
    completeness_scores = [metrics["completeness"] for metrics in metrics_over_time.values()]
    accuracy_scores = [metrics["accuracy"] for metrics in metrics_over_time.values()]
    diversity_scores = [metrics["diversity"] for metrics in metrics_over_time.values()]

    # Plot performance trends
    plt.figure(figsize=(10, 6))
    plt.plot(timestamps, relevance_scores, label="Relevance")
    plt.plot(timestamps, completeness_scores, label="Completeness")
    plt.plot(timestamps, accuracy_scores, label="Accuracy")
    plt.plot(timestamps, diversity_scores, label="Diversity")
    plt.title("LLM Performance Trends Over Time")
    plt.xlabel("Time")
    plt.ylabel("Score")
    plt.legend()
    plt.grid(True)
    plt.show()


Set Alerts for Anomalies
When performance drops below a threshold, trigger alerts to investigate:

In [ ]:
def monitor_performance(evaluation_results, thresholds):
    """
    Monitor performance metrics and trigger alerts if thresholds are breached.

    Args:
    - evaluation_results: Dictionary of evaluation scores.
    - thresholds: Dictionary defining acceptable thresholds for each metric.

    Returns:
    - List of alerts for metrics that fall below the threshold.
    """
    alerts = []
    for query, metrics in evaluation_results.items():
        for metric, score in metrics.items():
            if score < thresholds.get(metric, 0.5):  # Default threshold: 0.5
                alerts.append(f"Alert: {metric} for '{query}' below threshold with score {score:.2f}")
    return alerts

# Example Thresholds
thresholds = {
    "relevance": 0.7,
    "completeness": 0.6,
    "accuracy": 0.8,
    "diversity": 0.5
}

# Monitor performance and raise alerts
alerts = monitor_performance(evaluation_results, thresholds)
for alert in alerts:
    print(alert)


Simulate API Endpoints in Colab: Simulate API functionality using Python functions.

In [ ]:
!pip install flask


In [ ]:
from flask import Flask, request, jsonify
import threading
import requests

# Initialize Flask app
app = Flask(__name__)

# Define an example endpoint
@app.route('/recipe', methods=['POST'])
def get_recipe():
    """
    Simulate an API endpoint that takes user input and returns a recipe suggestion.
    """
    data = request.json  # Get JSON data from the request
    user_query = data.get('query', '')
    response = {
        "query": user_query,
        "recipes": [
            {"title": "Vegan Chocolate Cake", "ingredients": ["cocoa", "flour", "sugar"], "directions": "Mix and bake."},
            {"title": "Gluten-Free Pizza", "ingredients": ["almond flour", "cheese"], "directions": "Mix and bake."},
        ]
    }
    return jsonify(response)

# Run the Flask app in a separate thread
def run_app():
    app.run(host='0.0.0.0', port=5000)

thread = threading.Thread(target=run_app)
thread.daemon = True
thread.start()

# Test the endpoint using the requests library
import time
time.sleep(2)  # Wait for the Flask server to start

# Simulate a POST request to the API
test_data = {"query": "Suggest vegan recipes"}
response = requests.post("http://0.0.0.0:5000/recipe", json=test_data)
print(response.json())



In [ ]:
!pip install fastapi uvicorn


In [ ]:
from fastapi import FastAPI
from pydantic import BaseModel
import threading
import requests

# Initialize FastAPI app
app = FastAPI()

# Define request and response schemas
class RecipeRequest(BaseModel):
    query: str

class RecipeResponse(BaseModel):
    title: str
    ingredients: list
    directions: str

# Define an example endpoint
@app.post("/recipe", response_model=list[RecipeResponse])
async def get_recipe(request: RecipeRequest):
    """
    Simulate an API endpoint that takes user input and returns recipe suggestions.
    """
    recipes = [
        {"title": "Vegan Brownies", "ingredients": ["cocoa", "flour", "sugar"], "directions": "Mix and bake."},
        {"title": "Vegetable Stir Fry", "ingredients": ["broccoli", "carrot", "soy sauce"], "directions": "Stir fry in a pan."},
    ]
    return recipes

# Run the FastAPI app in a separate thread
import uvicorn
def run_app():
    uvicorn.run(app, host="0.0.0.0", port=8000)

thread = threading.Thread(target=run_app)
thread.daemon = True
thread.start()

# Test the endpoint using the requests library
import time
time.sleep(2)  # Wait for the FastAPI server to start

# Simulate a POST request to the API
test_data = {"query": "Suggest vegan recipes"}
response = requests.post("http://0.0.0.0:8000/recipe", json=test_data)
print(response.json())


In [ ]:
def simulate_recipe_api(user_query):
    """
    Simulate an API call by returning recipes based on user query.
    """
    response = {
        "query": user_query,
        "recipes": [
            {"title": "Vegan Pancakes", "ingredients": ["flour", "almond milk", "banana"], "directions": "Mix and cook."},
            {"title": "Gluten-Free Tacos", "ingredients": ["corn tortillas", "beans", "salsa"], "directions": "Assemble and serve."},
        ]
    }
    return response

# Simulate an API request
user_query = "Suggest recipes for breakfast"
api_response = simulate_recipe_api(user_query)
print(api_response)


Add Visualisation or Frontend: Use Colab widgets for interactivity.

In [ ]:
!pip install ipywidgets


In [ ]:
import ipywidgets as widgets
from IPython.display import display, clear_output
import pandas as pd
import numpy as np

# Sample dataset
recipes = pd.DataFrame({
    "title": ["Vegan Pancakes", "Gluten-Free Tacos", "Vegetable Stir Fry", "Chocolate Brownies", "Quinoa Salad"],
    "ingredients": [
        "flour, almond milk, banana",
        "corn tortillas, beans, salsa",
        "broccoli, carrot, soy sauce",
        "cocoa powder, sugar, flour",
        "quinoa, cucumber, tomato"
    ],
    "directions": [
        "Mix and cook on a griddle.",
        "Assemble and serve.",
        "Stir fry in a pan.",
        "Mix and bake in the oven.",
        "Combine and toss with dressing."
    ],
    "is_baking": [False, False, False, True, False]
})

# Widget for user query
query_widget = widgets.Text(
    value='',
    placeholder='Enter keywords (e.g., vegan, gluten-free)',
    description='Query:',
    layout=widgets.Layout(width='500px')
)

# Widget for non-baking prioritisation
non_baking_widget = widgets.Checkbox(
    value=True,
    description='Prioritize non-baking recipes',
    disabled=False
)

# Button to trigger recommendations
button = widgets.Button(description="Get Recommendations", button_style='success')

# Output widget for displaying results
output = widgets.Output()

def recommend_recipes(query, prioritize_non_baking=True, max_results=3):
    """
    Filter recipes based on the query and prioritisation.
    """
    # Basic keyword filtering
    filtered = recipes[recipes['ingredients'].str.contains(query, case=False, na=False)]

    if prioritize_non_baking:
        non_baking = filtered[~filtered['is_baking']]
        if len(non_baking) >= max_results:
            return non_baking.head(max_results)

    return filtered.head(max_results)

# Function to display recommendations
def on_button_click(b):
    with output:
        clear_output()  # Clear previous output
        user_query = query_widget.value
        prioritize_non_baking = non_baking_widget.value

        # Get recommendations
        recommendations = recommend_recipes(user_query, prioritize_non_baking)

        # Display results
        if recommendations.empty:
            print("No recipes found for your query.")
        else:
            for _, row in recommendations.iterrows():
                print(f"Title: {row['title']}")
                print(f"Ingredients: {row['ingredients']}")
                print(f"Directions: {row['directions']}\n")
                print("-" * 50)

# Link button to the function
button.on_click(on_button_click)

# Display the interface
display(query_widget, non_baking_widget, button, output)


In [ ]:
import matplotlib.pyplot as plt

def plot_recipe_types():
    counts = recipes['is_baking'].value_counts()
    labels = ['Non-Baking', 'Baking']
    plt.pie(counts, labels=labels, autopct='%1.1f%%', startangle=90)
    plt.title("Recipe Types")
    plt.show()

# Add a button to display the chart
chart_button = widgets.Button(description="Show Recipe Chart", button_style='info')

def on_chart_button_click(b):
    with output:
        clear_output()
        plot_recipe_types()

chart_button.on_click(on_chart_button_click)

# Display chart button
display(chart_button)


Save and Share Results: Allow users to download results or save the processed dataset.

In [ ]:
# Save the cleaned dataset as a new CSV
df_cleaned.to_csv('cleaned_recipes.csv', index=False)

# Provide a link to download
from google.colab import files
files.download('cleaned_recipes.csv')
